src/gui/reports/ReportsWidgetHibp.cpp

```diff
@@ -0,0 +1,331 @@
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "ReportsWidgetHibp.h"
+#include "ui_ReportsWidgetHibp.h"
+
+#include "core/Database.h"
+#include "core/Group.h"
+#include "gui/MessageBox.h"
+
+#include <QStandardItemModel>
+
+ReportsWidgetHibp::ReportsWidgetHibp(QWidget* parent)
+    : QWidget(parent)
+    , m_ui(new Ui::ReportsWidgetHibp())
+{
+    m_ui->setupUi(this);
+
+    m_referencesModel.reset(new QStandardItemModel());
+    m_ui->hibpTableView->setModel(m_referencesModel.data());
+    m_ui->hibpTableView->setSelectionMode(QAbstractItemView::NoSelection);
+    m_ui->hibpTableView->horizontalHeader()->setSectionResizeMode(QHeaderView::ResizeToContents);
+
+    connect(m_ui->hibpTableView, SIGNAL(doubleClicked(QModelIndex)), SLOT(emitEntryActivated(QModelIndex)));
+}
+
+ReportsWidgetHibp::~ReportsWidgetHibp()
+{
+}
+
+void ReportsWidgetHibp::loadSettings(QSharedPointer<Database> db)
+{
+    // Re-initialize
+    m_db = std::move(db);
+    m_checkStarted = false;
+    m_referencesModel->clear();
+    m_pwQueue.clear();
+    m_pwPwned.clear();
+    m_error.clear();
+    m_rowToEntry.clear();
+    m_edEntry = nullptr;
+    m_ui->progressBar->hide();
+
+    // Collect all passwords in the database (unless recycled, and
+    // unless empty).
+    for (const auto* group : m_db->rootGroup()->groupsRecursive(true)) {
+        if (!group->isRecycled()) {
+            for (const auto* entry : group->entries()) {
+                if (!entry->isRecycled() && !entry->password().isEmpty()) {
+                    m_pwQueue << entry->password();
+                }
+            }
+        }
+    }
+
+    // Store the number of passwords we need to check for the
+    // progress bar
+    m_qMax = m_pwQueue.size();
+}
+
+/*
+ * Fill the table will all entries that have passwords that we've
+ * found to have been pwned.
+ */
+void ReportsWidgetHibp::makeHibpTable()
+{
+
+    // Reset the table
+    m_referencesModel->clear();
+    m_referencesModel->setHorizontalHeaderLabels(QStringList()
+                                                 << tr("Title") << tr("Path") << tr("Password exposed …"));
+    m_rowToEntry.clear();
+
+    // Search database for passwords that we've found so far
+    std::vector<std::tuple<const Entry*, const Group*, int>> items;
+    for (const auto* group : m_db->rootGroup()->groupsRecursive(true)) {
+        if (!group->isRecycled()) {
+            for (const auto* entry : group->entries()) {
+                if (!entry->isRecycled()) {
+                    const auto found = m_pwPwned.find(entry->password());
+                    if (found != m_pwPwned.end()) {
+                        items.emplace_back(entry, group, *found);
+                    }
+                }
+            }
+        }
+    }
+
```

Comment:
You should be using standard Qt classes (QList, QPair). You do not need to store the group because you can access that from the entry: `entry->group()`.

Reply:
TIL, tx, that's gonna save quite some work.